keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,560 @@
+|Simple Timing wheels| circular buffer of MaxTimeOut slots. List of expiring timers at each slot. | Works for small bounded  MaxTimeOut which is not our case. Insertion and removal is O(1) via indexing |
+|Hashed Wheel| Hash expiring time and insert in a hash table with linked list at each index | Bookkeeping is O(1) and worst case insertion is O(n) |
+|Hierarchical Wheel| multiple timer wheels for different resolutions (Seconds, minutes, hours, days). When seconds rolls over we grab the next minutes timers and recreate the seconds wheel. similarly for minutes and hours. | Sharding at different hierarchy levels improves insertion and bookkeeping performance. |
+
+## Improvements/Considerations
+
+### Add Counters
+
+These counters would be added to `.status` section of the CronJob object:
+- `SuccessfulRuns` Count of all successful runs
+- `FailedRuns` Count of all failed runs
+- `FailuresSinceSuccess` Count of failed runs since last successful
+
+### Add .status.conditions
+
+Add a condition array with `Settled` condition type. This would help with the effort of standardizing conditions across all core types. `Settled` is set at the end of every successful reconcile run. The key thing to note here is the notions of `Settled` does not imply the `Job`s are running correctly. It just means that the controller is done processing this object successfully.
+
+NOTE: This should be handled across all of SIG-Apps owned controller.
+
+### Support Jitter for cronjobs
+
+We propose to introduce `.spec.jitter` which is a percentage of the time delta to the next schedule. We propose to cap it to 50%. There is also a [community request](https://github.com/kubernetes/community/issues/2440) for this.
+
+```golang
+delta = nextScheduleTime - currentTime
+jitter = delta*cronjob.Spec.Jitter/100
+nextScheduleTime += jitter
+```
+
+### Support Timezone for cronjobs
+
+We propose to introduce `.spec.timezone` which indicates the timezone to be considered when scheduling this cronjob.
+
+Clusters across different environments have different timezones. There is no known recommendation or conformance test to define timezone for the master. Different distributions may have different behavior. For cronjob, the timezone is inferred from the master. Someone with just API access has no idea what that controller manager timezone is. Sometimes cluster operators too, who run the controller manager in a container with a UTC timezone on a host with a non-UTC timezone.
+
+These conditions imply that people with just API access who want to predict when their job will run have no guarantee. By providing a timezone field we provide an absolute value instead of relying on the relative timezone of where the master is running. This ensures more portability of the CronJob configs across clusters as well.
+
+This is an optional field and when not present reverts to existing behavior of using the master's timezone.
```

Comment:
I wouldn't try to merge those two together. We need to graduate CronJob to GA to avoid it being removed. The admission controller is kind-of extension of the feature.

Reply:
I agree with Wojtek, it's pretty orthogonal to this proposal, it can be done but it's not a requirement.